src/DataStructures/DataBox/Item.hpp

```diff
@@ -0,0 +1,148 @@
+//
+// Its value may be fetched by calling db::get (which calls get)
+//
+// Its value may be changed by calling db::mutate (which calls mutate)
+template <typename Tag>
+class Item<Tag, Requires<db::is_mutable_item_tag_v<Tag>>> {
+ public:
+  static constexpr ItemType item_type = ItemType::Mutable;
+
+  using value_type = typename Tag::type;
+
+  constexpr Item() = default;
+  constexpr Item(Item const&) = default;
+  constexpr Item(Item&&) = default;
+  constexpr Item& operator=(Item const&) = default;
+  constexpr Item& operator=(Item&&) = default;
+  ~Item() = default;
+
+  explicit Item(value_type value) noexcept : value_(std::move(value)) {}
+
+  const value_type& get() const noexcept { return value_; }
+
+  value_type& mutate() noexcept { return value_; }
+
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pack_unpack(PUP::er& p) noexcept { p | value_; }
+
+ private:
+  value_type value_{};
+};
+
+// A compute item in a DataBox
+//
+// A compute item is an item in a DataBox whose value depends upon other items
+// in the DataBox.  It is lazily evaluated (i.e. not computed until it is
+// retrieved) and will be reevaluated if any of the items upon which it depends
+// is changed.
+//
+// A compute item is default constructed when the DataBox is constructed.
+//
+// When a compute item is fetched via db::get, db::get will check whether it has
+// already been evaluated (by calling evaluated).  If it has been evaluated,
+// db::get fetches its value (via get).  If it has not been evaluated (either
+// initially or after reset has been called), db::get will call evaluate which
+// will compute the value using tag::function.
+//
+// When db::mutate is called on a mutable item, all compute items that depend
+// (directly or indirectly) on the mutated item will have their reset function
+// called.
+//
+// A compute item may not be directly mutated (its value only changes after one
+// of its dependencies changes and it is fetched again)
+template <typename Tag>
+class Item<Tag, Requires<db::is_compute_tag_v<Tag>>> {
+ public:
+  static constexpr ItemType item_type = ItemType::Compute;
+
+  using value_type = typename Tag::type;
+
+  constexpr Item() = default;
+  constexpr Item(Item const&) = default;
+  constexpr Item(Item&&) = default;
+  constexpr Item& operator=(Item const&) = default;
+  constexpr Item& operator=(Item&&) = default;
+  ~Item() = default;
+
+  const value_type& get() const noexcept { return value_; }
+
+  bool evaluated() const noexcept { return evaluated_; }
+
+  void reset() noexcept { evaluated_ = false; }
+
+  template <typename... Args>
+  void evaluate(const Args&... args) const noexcept {
+    Tag::function(make_not_null(&value_), args...);
+    evaluated_ = true;
+  }
+
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pack_unpack(PUP::er& p) noexcept {
+    p | evaluated_;
+    if (evaluated_) {
+      p | value_;
+    }
+  }
+
+ private:
+  mutable value_type value_{};
+  mutable bool evaluated_{false};
```

Comment:
Can this be a `std::optional`?

Reply:
I thought of that, but that would introduce a reallocation every time the compute is reset

Reply:
Could you add a comment, so we don't accidentally make this an optional in the future? It's absolutely fine to do this in a follow-up PR to avoid holding up this one if you prefer.